File: codeforces/ac/014/D.hs
import Data.Array (accumArray, (!))
import Data.List (sort)

gao n e = maximum [fst (gao' a b) * fst (gao' b a) | (a, b) <- e] where
        e' = accumArray (flip (:)) [] (1, n) $ concat [e, [(b, a) | (a, b) <- e]]
        gao' p q = (\(a, b) -> let b' = reverse $ sort b in (max a $ (b'!!0)+(b'!!1), b'!!0)) $
                foldl (\(a,b) (c,d) -> (max a c, d+1:b)) (0, [0, 0]) [gao' i p | i <- e'!p, i /= q]

main = do
        v <- getLine
        e <- getContents
        putStrLn $ show $ let n = read v in gao n $ map ((\[a,b]->(a,b)) . map read . words) $ lines e
{-
WA25!! max a (b'!!0)+(b'!!1) => max a $ (b'!!0)+(b'!!1)
# 	When 	Who 	Problem 	Lang 	Verdict 	Time 	Memory
234865 	Dec 29, 2010 1:58:01 PM 	watashi 	D - Two Paths 	Haskell 	Accepted 	50 ms 	1548 KB
-}

File: codeforces/ac/014/D.rb
def gao(e, p, q)
	l, t = 0, [0, 0]
	e[p].each do |i|
		next if i == q
		ll, dd = gao(e, i, p)
		l = [l, ll].max
		t <<= dd + 1
	end
	t.sort!{|a, b| b <=> a}
	return [[l, t[0] + t[1]].max, t[0]]
end

n = gets.to_i
e = Array.new(n).map{[]}
ee = []
(n - 1).times do
	a, b = gets.split.map{|i| i.to_i - 1}
	e[a] <<= b
	e[b] <<= a
	ee <<= [a, b]
end
p ee.map{|p| gao(e, p[0], p[1])[0] * gao(e, p[1], p[0])[0]}.max

__END__
# 	When 	Who 	Problem 	Lang 	Verdict 	Time 	Memory
212753 	Dec 9, 2010 2:17:30 PM 	watashi 	D - Two Paths 	Ruby 	Accepted 	310 ms 	3096 KB
